revscoring/scorer_models/thresholds.py

```diff
@@ -0,0 +1,167 @@
+                     max_thresholds):
+        """
+        Builds a set of threshold stats for a single label.
+        """
+        probas = (s['probability'][label] for s, l in score_labels)
+        thresholds = Thresholds._limit_thresholds(probas, max_thresholds)
+
+        threshold_stats = []
+        for threshold in thresholds:
+            y_preds = [s['probability'][label] >= threshold
+                       for s, l in score_labels]
+            y_trues = [l == label for s, l in score_labels]
+
+            # Generate counts
+            true_positives = 0
+            false_positives = 0
+            true_negatives = 0
+            false_negatives = 0
+            for y_pred, y in zip(y_preds, y_trues):
+                true_positives += y_pred and y
+                false_positives += y_pred and not y
+                true_negatives += not y_pred and not y
+                false_negatives += not y_pred and y
+
+            # Apply scaling
+            true_positives = true_positives / label_rate
+            false_negatives = false_negatives / label_rate
+            true_negatives = true_negatives / non_label_rate
+            false_positives = false_positives / non_label_rate
+
+            # Useful variables
+            n = true_positives + true_negatives + \
+                false_positives + false_negatives
+            positives = true_positives + false_positives
+            negatives = true_negatives + false_negatives
+            trues = true_positives + false_negatives
+            falses = false_positives + true_negatives
+            true_predictions = true_positives + true_negatives
+
+            # Generate stats
+            accuracy = true_predictions / n
+            recall = true_positives / max(trues, 1)
+            _recall = true_negatives / max(falses, 1)
+            precision = true_positives / max(positives, 1)
+            _precision = true_negatives / max(negatives, 1)
+            f1 = (precision * recall) / max(precision + recall, 1)
+            _f1 = (_precision * _recall) / max(_precision + _recall, 1)
+
+            # Build the record
+            threshold_stats.append({
+                'threshold': threshold,
+                'accuracy': accuracy,
+                'recall': recall,
+                '!recall': _recall,
+                'precision': precision,
+                '!precision': _precision,
+                'f1': f1,
+                '!f1': _f1
+            })
+        return threshold_stats
+
+    @staticmethod
+    def _limit_thresholds(decision, max_thresholds):
+        """
+        Limits the range of a decision function (usually a probability) to
+        a representative set of thresholds.
+        """
+        unique_thresholds = sorted(set(decision))
+        if len(unique_thresholds) < max_thresholds:
+            thresholds = unique_thresholds
+        else:
+            step = len(unique_thresholds) / max_thresholds
```

Comment:
Ooh that gets messy.  But I guess you wanted to spread out the sampling error, rather than just unique_thresholds[::step].